File: tps/tp03/TP03Q03/TP03Q03.java
import java.io.*;
import java.lang.Math;

class Celula {
    public Personagem elemento; // Elemento inserido na Celula
    public Celula prox; // Aponta a Celula -> Prox

    public Celula() {
        this(new Personagem());
    }

    public Celula(Personagem elemento) {
        this.elemento = elemento;
        this.prox = null;
    }
}

class Personagem {
    private String nome;
    private int altura;
    private double peso;
    private String corDoCabelo;
    private String corDaPele;
    private String corDosOlhos;
    private String anoNascimento;
    private String genero;
    private String homeworld;

    public Personagem() {
    }

    public Personagem(String nome, int altura, double peso, String corDoCabelo, String corDaPele,
            String corDosOlhos, String anoNascimento, String genero, String homeworld) {
        this.nome = nome;
        this.altura = altura;
        this.peso = peso;
        this.corDoCabelo = corDoCabelo;
        this.corDaPele = corDaPele;
        this.corDosOlhos = corDosOlhos;
        this.anoNascimento = anoNascimento;
        this.genero = genero;
        this.homeworld = homeworld;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getNome() {
        return nome;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public int getAltura() {
        return altura;
    }

    public void setPeso(double peso) {
        this.peso = peso;
    }

    public double getPeso() {
        return peso;
    }

    public void setCorDoCabelo(String corDoCabelo) {
        this.corDoCabelo = corDoCabelo;
    }

    public String getCorDoCabelo() {
        return corDoCabelo;
    }

    public void setCorDaPele(String corDaPele) {
        this.corDaPele = corDaPele;
    }

    public String getCorDaPele() {
        return corDaPele;
    }

    public void setCorDosOlhos(String corDosOlhos) {
        this.corDosOlhos = corDosOlhos;
    }

    public String getCorDosOlhos() {
        return corDosOlhos;
    }

    public String getAnoNascimento() {
        return anoNascimento;
    }

    public void setAnoNascimento(String anoNascimento) {
        this.anoNascimento = anoNascimento;
    }

    public String getGenero() {
        return genero;
    }

    public void setGenero(String genero) {
        this.genero = genero;
    }

    public String getHomeworld() {
        return homeworld;
    }

    public void setHomeworld(String homeworld) {
        this.homeworld = homeworld;
    }

    public Personagem clone() {
        Personagem cloned = new Personagem();

        cloned.nome = this.nome;
        cloned.altura = this.altura;
        cloned.peso = this.peso;
        cloned.corDoCabelo = this.corDoCabelo;
        cloned.corDaPele = this.corDaPele;
        cloned.corDosOlhos = this.corDosOlhos;
        cloned.anoNascimento = this.anoNascimento;
        cloned.genero = this.genero;
        cloned.homeworld = this.homeworld;

        return cloned;
    }
}

class Fila {
    private Celula primeiro;
    private Celula ultimo;

    int tamanhoFila = 6; // Exercício exigia fila de tamanho 5
    int primeiroContador = 0, ultimoContador = 0;

    public Fila() {
        primeiro = new Celula();
        ultimo = primeiro;
    }

    public void inserir(Personagem personagem) throws Exception {
        if (((ultimoContador + 1) % tamanhoFila) == primeiroContador) {
            remover(); // Remover item caso a fila esteja cheia
        }

        ultimo.prox = new Celula(personagem);
        ultimo = ultimo.prox;

        ultimoContador = (ultimoContador + 1) % tamanhoFila; // Análise de acordo com a Fila Circular padrão

        calculaMedia(); // Calcula a média das alturas
    }

    public Personagem remover() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover!");
        }

        Celula tmp = primeiro;
        primeiro = primeiro.prox;
        Personagem resp = primeiro.elemento;
        tmp.prox = null;
        tmp = null;

        primeiroContador = (primeiroContador + 1) % tamanhoFila;

        return resp;
    }

    // Exibir a média arredondada das alturas
    public void calculaMedia() {
        int contador = 0;
        double soma = 0;

        for (Celula i = primeiro.prox; i != null; i = i.prox) {
            soma += i.elemento.getAltura();
            contador++;
        }

        MyIO.println((int) (Math.round(soma / contador))); // Garantir o arredondamento necessário
    }
}

public class TP03Q03 {
    public static void main(String[] args) throws Exception {
        MyIO.setCharset("ISO-8859-1");

        Fila listaPersonagens = new Fila();

        String caminhoArquivo = MyIO.readLine().replaceAll("é", "\u00e9");

        // Testar fim da primeira parte do arquivo
        while (testaFim(caminhoArquivo) == false) {

            // Montar personagem e adicionar ao fim da lista
            listaPersonagens.inserir(montaPersonagem(caminhoArquivo));

            caminhoArquivo = MyIO.readLine().replaceAll("é", "\u00e9");
        }

        int quantidadeRegistros = MyIO.readInt(); // Ler quantidade de registros a serem inseridos ou removidos

        analisaOperacoes(quantidadeRegistros, listaPersonagens);
    }

    // Testar fim do arquivo
    public static boolean testaFim(String palavra) {
        boolean teste = false;

        if (palavra.equals("FIM")) {
            teste = true;
        }

        return teste;
    }

    // Retornar personagem lido
    public static Personagem montaPersonagem(String caminhoArquivo) throws Exception {
        RandomAccessFile leitura = new RandomAccessFile(caminhoArquivo, "r");

        String descricaoPersonagem = leitura.readLine().replaceAll("é", "\u00e9");

        Personagem personagem = new Personagem();
        int contadorDoisPontos = 0; // Contar separadores

        for (int i = 0; i < descricaoPersonagem.length(); i++) {
            if (descricaoPersonagem.charAt(i) == ':') {
                contadorDoisPontos++;

                switch (contadorDoisPontos) {
                    case 1:
                        personagem.setNome(leituraAtributo(descricaoPersonagem, i + 3));
                        break;
                    case 2:
                        String atributoInteiro = leituraAtributo(descricaoPersonagem, i + 3);
                        if (atributoInteiro.equals("unknown")) {
                            personagem.setAltura(0);
                        } else {
                            personagem.setAltura(Integer.parseInt(atributoInteiro));
                        }
                        break;
                    case 3:
                        String atributoDouble = leituraAtributo(descricaoPersonagem, i + 3).replaceAll(",", ".");
                        if (atributoDouble.equals("unknown")) {
                            personagem.setPeso(0);
                        } else {
                            personagem.setPeso(Double.parseDouble(atributoDouble));
                        }
                        break;
                    case 4:
                        personagem.setCorDoCabelo(leituraAtributo(descricaoPersonagem, i + 3));
                        break;
                    case 5:
                        personagem.setCorDaPele(leituraAtributo(descricaoPersonagem, i + 3));
                        break;
                    case 6:
                        personagem.setCorDosOlhos(leituraAtributo(descricaoPersonagem, i + 3));
                        break;
                    case 7:
                        personagem.setAnoNascimento(leituraAtributo(descricaoPersonagem, i + 3));
                        break;
                    case 8:
                        personagem.setGenero(leituraAtributo(descricaoPersonagem, i + 3));
                        break;
                    case 9:
                        personagem.setHomeworld(leituraAtributo(descricaoPersonagem, i + 3));
                        i = descricaoPersonagem.length(); // Encerra os ciclos de repetição desnecessários
                        break;
                    default:
                        break;
                }
            }
        }

        leitura.close();

        return personagem;
    }

    // Retornar valor do atributo do personagem
    public static String leituraAtributo(String descricaoPersonagem, int index) {
        String atributo = new String();

        while (descricaoPersonagem.charAt(index) != '\'') {
            atributo += descricaoPersonagem.charAt(index);

            index++;
        }

        return atributo;
    }

    // Analisar manipulações que devem ser feitas nos registros
    public static void analisaOperacoes(int quantidadeRegistros, Fila listaPersonagens) throws Exception {

        for (int i = 0; i < quantidadeRegistros; i++) {
            String comandoOperacao = MyIO.readString();
            String caminhoOperacao = new String();

            // Analisar o comando de cada operação a ser executada
            switch (comandoOperacao) {
                case "I":
                    caminhoOperacao = MyIO.readString().replaceAll("é", "\u00e9");
                    listaPersonagens.inserir(montaPersonagem(caminhoOperacao));
                    break;
                case "R":
                    MyIO.println("(R) " + listaPersonagens.remover().getNome());
                    break;
                default:
                    i = quantidadeRegistros; // Redundância para garantir o fim
                    break;
            }
        }
    }
}

File: tps/tp03/TP03Q03/TP03Q03.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

// Definição do registro do personagem
typedef struct Personagem
{
    char nome[40];
    int altura;
    double peso;
    char corDoCabelo[40];
    char corDaPele[40];
    char corDosOlhos[40];
    char anoNascimento[40];
    char genero[40];
    char homeworld[40];
} Personagem;

// Função para retornar o resto da divisão com double
double fmod(double x, double y)
{
    return x - (int)(x / y) * y;
}

// Capturar o atributo entre aspas simples
void leituraAtributo(char atributo[], char descricaoPersonagem[], int index)
{
    int i = 0;

    while (descricaoPersonagem[index] != '\'')
    {
        atributo[i] = descricaoPersonagem[index];

        i++;
        index++;
    }

    atributo[i] = '\0';
}

// Função para testar o fim do arquivo
bool testaFim(char palavra[])
{
    bool teste = false;

    if (palavra[0] == 'F' && palavra[1] == 'I' && palavra[2] == 'M')
    {
        teste = true;
    }

    return teste;
}

// Função para estruturar o personagem
Personagem montaPersonagem(char caminhoArquivo[])
{
    FILE *leitura = fopen(caminhoArquivo, "r");

    char descricaoPersonagem[1000];

    fscanf(leitura, " %[^\n]s", descricaoPersonagem);

    Personagem personagem; // Cria a variável struct

    int contador = 0;

    for (int i = 0; i < strlen(descricaoPersonagem); i++)
    {
        if (descricaoPersonagem[i] == ':')
        {
            char atributo[50];
            contador++;

            switch (contador)
            {
            case 1:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.nome, atributo);
                break;
            case 2:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                personagem.altura = atoi(atributo);
                break;
            case 3:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                for (int i = 0; i < strlen(atributo); i++)
                {
                    if (atributo[i] == ',')
                    {
                        atributo[i] = atributo[i - 1];
                        atributo[i - 1] = '0';
                    }
                }
                personagem.peso = atof(atributo);
                break;
            case 4:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.corDoCabelo, atributo);
                break;
            case 5:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.corDaPele, atributo);
                break;
            case 6:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.corDosOlhos, atributo);
                break;
            case 7:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.anoNascimento, atributo);
                break;
            case 8:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.genero, atributo);
                break;
            case 9:
                leituraAtributo(atributo, descricaoPersonagem, i + 3);
                strcpy(personagem.homeworld, atributo);

                i = strlen(descricaoPersonagem); // Encerra os ciclos de repetição desnecessários
                break;
            default:
                break;
            }
        }
    }

    fclose(leitura);

    return personagem;
}

#define TAMANHOFILA 6 // Exercício exigia fila circular de cinco elementos

int primeiroContador = 0, ultimoContador = 0; // Controlar o tamanho da Fila Circular por contadores

// Definição da Celula
typedef struct Celula
{
    Personagem elemento; // Elemento inserido na Celula
    struct Celula *prox; // Aponta para Celula -> Prox
} Celula;

Celula *novaCelula(Personagem elemento)
{
    Celula *nova = (Celula *)malloc(sizeof(Celula));
    nova->elemento = elemento;
    nova->prox = NULL;

    return nova;
}

// Criação da Fila
Celula *primeiro;
Celula *ultimo;

// Cria uma fila sem elementos
void start()
{
    Personagem personagem;

    primeiro = novaCelula(personagem);
    ultimo = primeiro;
}

// Exibir a média arredondada das alturas
void calculaMedia()
{
    int contador = 0, media;
    double soma = 0;

    for (Celula *i = primeiro->prox; i != NULL; i = i->prox)
    {
        soma += i->elemento.altura;
        contador++;
    }

    media = round(soma / contador);

    printf("%d\n", media);
}

// Remover elemento da fila (FIFO)
Personagem remover()
{
    if (primeiro == ultimo)
    {
        printf("Erro ao remover!");
    }
    Celula *tmp = primeiro;
    primeiro = primeiro->prox;
    Personagem resp = primeiro->elemento;
    tmp->prox = NULL;
    free(tmp);
    tmp = NULL;

    primeiroContador = (primeiroContador + 1) % TAMANHOFILA;

    return resp;
}

// Inserir elemento na fila (FIFO)
void inserir(Personagem personagem)
{
    if (((ultimoContador + 1) % TAMANHOFILA) == primeiroContador)
    {
        remover(); // Remover item caso a Fila esteja cheia
    }

    ultimo->prox = novaCelula(personagem);
    ultimo = ultimo->prox;

    ultimoContador = (ultimoContador + 1) % TAMANHOFILA; // Análise de acordo com a Fila Circular padrão

    calculaMedia(); // Calcula a média das alturas
}

// Exibir o resultado
void imprimeResultado()
{
    Celula *i;
    for (i = primeiro->prox; i != NULL; i = i->prox)
    {
        printf(" ## %s", i->elemento.nome);
        printf(" ## %d", i->elemento.altura);
        if (fmod(i->elemento.peso, 1) == 0)
            printf(" ## %.0lf", i->elemento.peso);
        else
            printf(" ## %.1lf", i->elemento.peso);
        printf(" ## %s", i->elemento.corDoCabelo);
        printf(" ## %s", i->elemento.corDaPele);
        printf(" ## %s", i->elemento.corDosOlhos);
        printf(" ## %s", i->elemento.anoNascimento);
        printf(" ## %s", i->elemento.genero);
        printf(" ## %s", i->elemento.homeworld);
        printf(" ## \n");
    }
}

int main(void)
{
    char caminhoArquivo[100], nomePersonagem[100];
    int contadorTamanho = 0;
    start();

    scanf(" %[^\n]s", caminhoArquivo);
    getchar();

    // Testa o fim da primeira parte do arquivo
    while (testaFim(caminhoArquivo) == false)
    {
        inserir(montaPersonagem(caminhoArquivo));
        contadorTamanho += 1;

        scanf(" %[^\n]s", caminhoArquivo);
        getchar();
    }

    int quantidadeRegistros;
    scanf("%d", &quantidadeRegistros);

    // Analisar manipulações que devem ser feitas nos registros
    for (int i = 0; i < quantidadeRegistros; i++)
    {
        char comandoOperacao[5];
        scanf(" %s", comandoOperacao);
        getchar();

        char caminhoOperacao[100];

        // Analisar o comando de cada operação a ser executada
        if (comandoOperacao[0] == 'I') // Casos de inserção
        {
            scanf(" %[^\n]s", caminhoOperacao);
            getchar();

            inserir(montaPersonagem(caminhoOperacao));
        }
        else if (comandoOperacao[0] == 'R') // Casos de remoção
        {
            printf("(R) %s\n", remover().nome);
        }
    }

    return 0;
}